File: ai/src/retrieval.py
"""Turn the published portfolio bundle into the few paragraphs one question needs.

Deterministic, no embeddings. The portfolio is a few dozen records; a token-overlap score with
a category prior finds the right ones, costs nothing, and — unlike a vector index — cannot
drift out of date, because it is recomputed from the bundle on every request. Add vector
search only if real questions start missing.

The bundle arrives as decoded JSON, so everything here reads defensively: a field the CMS has
not filled in yet is absent, not empty, and must not raise.
"""

from __future__ import annotations

from urllib.parse import urlparse

from models import Doc, Source

#: Tokens that carry no signal but appear in most questions. Deliberately short — an
#: aggressive stoplist starts eating words like "work" and "use" that do discriminate here.
STOPWORDS = frozenset(
    """
    a an and are as at be by can did do does for from has have he her him his how i in is it
    its me my of on or she that the their them there they this to was were what when where
    which who whom whose why will with you your sumit sumits tell show about please
    """.split()
)

#: Question words that point at a kind of record. A hit is a prior, not a filter: it lifts a
#: whole category so "what has he built?" reaches the projects even though no project contains
#: the word "built", while still letting a strong term match in another kind outrank it.
ROUTES: tuple[tuple[tuple[str, ...], tuple[str, ...]], ...] = (
    (("project", "projects", "built", "build", "made", "portfolio", "demo", "app", "shipped"), ("project",)),
    (("skill", "skills", "tech", "technology", "technologies", "stack", "language", "languages", "tools", "know"), ("skill",)),
    (("experience", "work", "worked", "job", "role", "intern", "internship", "company"), ("experience",)),
    (("education", "study", "studying", "studied", "degree", "college", "university", "school", "gpa"), ("education",)),
    (("certificate", "certificates", "certification", "certifications", "certified", "course", "courses"), ("certificate",)),
    (("contact", "email", "reach", "github", "linkedin", "hire", "link", "links", "profile", "social"), ("link", "site")),
    (("who", "background", "bio", "summary", "introduce", "himself", "overview"), ("site",)),
)

#: How many documents may reach the model, and the total characters they may occupy. Both are
#: about cost and focus, not about a context-window limit: a shorter, relevant context is what
#: keeps the model from wandering into a neighbouring project.
TOP_K = 6
CONTEXT_BUDGET = 6000

_KIND_LABEL = {
    "project": "Project",
    "skill": "Skills",
    "experience": "Experience",
    "education": "Education",
    "certificate": "Certificate",
    "link": "Link",
    "site": "About",
}


def tokens(text: str) -> set[str]:
    """Lowercase word tokens, keeping the characters that make technology names distinct.

    `+`, `#` and `.` stay inside a token so "c++", "c#" and "scikit-learn.js" survive as
    themselves rather than collapsing into "c" and matching every sentence in the portfolio.
    """
    out: set[str] = set()
    word: list[str] = []
    for char in text.lower():
        if char.isalnum() or char in "+#.":
            word.append(char)
        elif word:
            out.add("".join(word).strip("."))
            word = []
    if word:
        out.add("".join(word).strip("."))
    return {w for w in out if w}


def _text(value: object) -> str:
    return value if isinstance(value, str) else ""


def _list(value: object) -> list:
    return value if isinstance(value, list) else []


def _section_text(section: object) -> str:
    """Flatten one project section — prose, flow diagram or metric table — into a sentence."""
    if not isinstance(section, dict):
        return ""
    body = section.get("body")
    body = body if isinstance(body, dict) else {}
    parts = [_text(section.get("heading")), _text(body.get("text"))]
    for row in _list(body.get("flow")) + _list(body.get("metrics")):
        parts.extend(_text(cell) for cell in _list(row))
    return " ".join(p for p in parts if p)


def _host(url: str) -> str:
    try:
        return (urlparse(url).hostname or "").removeprefix("www.")
    except ValueError:
        return ""


def _doc(kind: str, slug: str, title: str, text: str, extra: str = "") -> Doc:
    return Doc(
        kind=kind,
        slug=slug,
        title=title,
        text=text,
        terms=frozenset(tokens(f"{title} {text} {extra}")),
    )


def _project_doc(raw: object) -> Doc | None:
    if not isinstance(raw, dict):
        return None
    title = _text(raw.get("title"))
    if not title:
        return None
    stack = [_text(s) for s in _list(raw.get("stack")) if _text(s)]
    links = [raw_link for raw_link in _list(raw.get("links")) if isinstance(raw_link, dict)]
    lines = [
        f"{title} — {_text(raw.get('tagline'))}".strip(" —"),
        f"Status: {_text((raw.get('status') or {}).get('label')) if isinstance(raw.get('status'), dict) else ''}".strip(),
        f"Stack: {', '.join(stack)}" if stack else "",
        *(_section_text(s) for s in _list(raw.get("sections"))),
        _text(raw.get("note")),
        _text(raw.get("caveat")),
        *(f"{_text(link.get('label'))}: {_host(_text(link.get('url')))}" for link in links),
    ]
    aliases = " ".join(_text(a) for a in _list(raw.get("aliases")))
    return _doc(
        "project",
        _text(raw.get("slug")),
        title,
        "\n".join(line for line in lines if line and line != "Status:"),
        extra=f"{aliases} {_text(raw.get('desktopLabel'))}",
    )


def flatten(content: object) -> list[Doc]:
    """The published bundle as a flat list of retrievable documents.

    Every record the CMS publishes becomes exactly one Doc. Nothing is filtered here: the
    caller was handed a bundle that D1 already restricted to `published = 1`, so there is no
    second visibility rule to get wrong.
    """
    content = content if isinstance(content, dict) else {}
    docs: list[Doc] = []

    site = content.get("site")
    if isinstance(site, dict):
        name = _text(site.get("name"))
        paragraphs = " ".join(_text(p) for p in _list(site.get("paragraphs")))
        body = " ".join(p for p in [_text(site.get("subtitle")), paragraphs] if p)
        if body:
            docs.append(_doc("site", "", name or "About", body))

    for raw in _list(content.get("projects")):
        doc = _project_doc(raw)
        if doc:
            docs.append(doc)

    for raw in _list(content.get("skills")):
        if not isinstance(raw, dict):
            continue
        items = [_text(i) for i in _list(raw.get("items")) if _text(i)]
        heading = _text(raw.get("heading"))
        if items:
            docs.append(_doc("skill", "", heading or "Skills", f"{heading}: {', '.join(items)}"))

    for kind in ("experience", "education"):
        for raw in _list(content.get(kind)):
            if not isinstance(raw, dict):
                continue
            title = _text(raw.get("title"))
            if not title:
                continue
            body = " · ".join(p for p in [_text(raw.get("detail")), _text(raw.get("hint"))] if p)
            docs.append(_doc(kind, "", title, f"{title} — {body}".strip(" —")))

    for raw in _list(content.get("certificates")):
        if not isinstance(raw, dict):
            continue
        title = _text(raw.get("title"))
        if not title:
            continue
        issued = _text(raw.get("issueDate"))
        body = f"{title} — issued by {_text(raw.get('issuer'))}"
        docs.append(_doc("certificate", "", title, f"{body} ({issued})" if issued else body))

    for raw in _list(content.get("socialLinks")):
        if not isinstance(raw, dict):
            continue
        label = _text(raw.get("label"))
        url = _text(raw.get("url"))
        if not label or not url:
            continue
        handle = _text(raw.get("handle"))
        docs.append(_doc("link", "", label, f"{label}: {url}" + (f" ({handle})" if handle else ""), extra=_host(url)))

    return docs


def route(question: str) -> set[str]:
    """The kinds a question is asking about, empty when it names none."""
    words = tokens(question)
    kinds: set[str] = set()
    for triggers, targets in ROUTES:
        if words.intersection(triggers):
            kinds.update(targets)
    return kinds


def score(question: str, doc: Doc) -> int:
    """How well one document's own words answer one question, ignoring its category.

    Three signals, weighted by how much each one actually means. A term appearing anywhere in
    a record is weak evidence; the same term in its title is strong; the record's whole name
    quoted in the question is near-conclusive, which is what makes "tell me about the PM2.5
    forecasting project" land on that project rather than on every record mentioning PM2.5.
    """
    asked = tokens(question) - STOPWORDS
    if not asked:
        return 0

    title_terms = tokens(doc.title)
    total = 0
    for term in asked:
        if term in doc.terms:
            total += 2
        if term in title_terms:
            total += 3

    name = doc.title.lower()
    if len(name) > 3 and name in question.lower():
        total += 5
    return total


#: What a category prior is worth. Below a title hit and below a name quote, so routing orders
#: results that words already found and never overturns them.
ROUTE_BONUS = 2


def retrieve(question: str, docs: list[Doc], limit: int = TOP_K) -> list[Doc]:
    """The documents worth showing the model, best first, within the character budget.

    A category can be read two ways and the difference is what makes this work. "What projects
    has Sumit built?" names a category and nothing else, so every project is an answer.
    "Which projects use PyTorch?" names a category *and* a discriminator, so the category is
    only a hint and the discriminator decides — returning all three projects there would be
    the wrong answer dressed up as thoroughness.

    So: if the routed category contains a record the question's own words matched, the words
    win and routing merely reorders. If it contains none, the question is a browse and the
    whole category comes back.

    An empty result is a real answer, not a failure: the question is about something the
    portfolio does not cover, and the caller says so rather than asking the model to improvise.
    """
    kinds = route(question)
    scored = [(score(question, doc), i, doc) for i, doc in enumerate(docs)]
    browsing = bool(kinds) and not any(s > 0 for s, _, doc in scored if doc.kind in kinds)

    ranked = [
        (s + ROUTE_BONUS if doc.kind in kinds and (s > 0 or browsing) else s, i, doc)
        for s, i, doc in scored
    ]
    ranked = sorted((r for r in ranked if r[0] > 0), key=lambda r: (-r[0], r[1]))

    picked: list[Doc] = []
    budget = CONTEXT_BUDGET
    for _, _, doc in ranked[:limit]:
        cost = len(doc.text)
        if picked and cost > budget:
            break
        picked.append(doc)
        budget -= cost
    return picked


def sources(docs: list[Doc]) -> tuple[Source, ...]:
    """Citations for the UI. Only kinds with a public page carry a slug to link to."""
    seen: set[tuple[str, str]] = set()
    out: list[Source] = []
    for doc in docs:
        key = (doc.kind, doc.title)
        if key in seen:
            continue
        seen.add(key)
        out.append(Source(type=doc.kind, slug=doc.slug if doc.kind == "project" else "", title=doc.title))
    return tuple(out)


def as_context(docs: list[Doc]) -> str:
    """The retrieved documents as the plain text block the prompt embeds."""
    return "\n\n".join(f"[{_KIND_LABEL.get(doc.kind, doc.kind)}] {doc.text}" for doc in docs)

File: worker/mcp/retrieval.ts
// Published bundle → searchable documents → a small, ranked, bounded answer. Pure: no I/O.
//
// Deterministic on purpose, the same call as ai/src/retrieval.py makes for Ask Sumit: the corpus
// is a few dozen records, a token-overlap score finds the right ones, and an index recomputed
// from the bundle on every request cannot drift out of date. A project published in /admin is
// searchable here as soon as the bundle cache turns over — there is nothing to rebuild.
//
// ponytail: token overlap with a coverage floor. Add embeddings only when real questions miss.

import type { Content, ProjectSection } from '../../src/data/content'
import type { Doc, DocType, Source } from './types'
import { NO_MATCH } from './types'

export const MAX_RESULTS = 8
export const SNIPPET_CHARS = 280
export const MAX_TEXT_CHARS = 6000

/** Words that carry no signal but turn up in most questions. Short, like the Python list. */
const STOPWORDS = new Set(
  `a an and are as at be by can did do does for from has have he her him his how i in is it its me
  my of on or she that the their them there they this to was were what when where which who whom
  whose why will with you your sumit sumits tell show about please give list explain describe used
  use`.split(/\s+/),
)

/**
 * Question words that point at a kind of record. They count as a hit on every doc of that kind,
 * so "what are his projects?" reaches the projects even though no project contains "projects".
 */
const ROUTES: [string[], DocType[]][] = [
  [['project', 'built', 'build', 'made', 'portfolio', 'demo', 'app', 'shipped', 'work'], ['project']],
  [['research', 'study', 'paper', 'experiment'], ['research', 'project']],
  [['skill', 'tech', 'technology', 'stack', 'language', 'tool', 'know'], ['skill']],
  [['experience', 'job', 'role', 'intern', 'internship', 'company', 'worked'], ['experience']],
  [['education', 'degree', 'college', 'university', 'school', 'studied', 'gpa'], ['education']],
  [['certificate', 'certification', 'certified', 'course'], ['certificate']],
  [['resume', 'cv'], ['resume']],
  [['contact', 'email', 'reach', 'github', 'linkedin', 'hire', 'link', 'profile', 'social', 'bio', 'background', 'who'], ['profile']],
]

/**
 * What a caller means by a section name, against the headings projects actually carry. Also
 * used to widen a search term, so "methodology" finds a section headed "Forecasting pipeline".
 */
export const SECTION_SYNONYMS: Record<string, string[]> = {
  overview: ['what it is', 'what it does', 'overview', 'summary', 'about'],
  methodology: ['methodology', 'method', 'approach', 'pipeline', 'formula', 'model', 'implementation', 'how it works'],
  architecture: ['architecture', 'system', 'design', 'model', 'stack'],
  dataset: ['dataset', 'data', 'input features', 'features', 'observations'],
  experiments: ['experiment', 'validation', 'evaluation', 'testing'],
  results: ['results', 'result', 'performance', 'metrics', 'evaluation', 'validation', 'finding'],
  metrics: ['metrics', 'performance', 'results', 'evaluation'],
  deployment: ['deployment', 'api', 'demo', 'serving'],
  limitations: ['limitation', 'caveat', 'note'],
  references: ['links', 'references', 'source'],
}

// ---- text helpers ---------------------------------------------------------------------------

/**
 * Lowercase word tokens. `+` and `#` stay inside a token so "c++" and "c#" survive as themselves;
 * a trailing plural `s` is dropped so "projects" meets "project". Single letters are noise.
 */
export function tokens(text: string): Set<string> {
  const out = new Set<string>()
  for (const raw of text.toLowerCase().split(/[^a-z0-9+#.]+/)) {
    let word = raw.replace(/^\.+|\.+$/g, '')
    if (word.length > 3 && word.endsWith('s') && !word.endsWith('ss')) word = word.slice(0, -1)
    if (word.length > 1 || /[+#]/.test(word)) out.add(word)
  }
  return out
}

export const slugify = (text: string): string =>
  text
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
    .slice(0, 80)

function sectionText(section: ProjectSection): string {
  const body = section.body as Record<string, unknown>
  if (typeof body.text === 'string') return body.text
  if (Array.isArray(body.flow)) {
    return (body.flow as unknown[][]).map((row) => row.filter(Boolean).join(' — ')).join('\n')
  }
  if (Array.isArray(body.metrics)) {
    return (body.metrics as unknown[][])
      .map(([label, value, hint]) => `${label}: ${value}${hint ? ` (${hint})` : ''}`)
      .join('\n')
  }
  return ''
}

const lines = (...parts: (string | undefined | false)[]) => parts.filter(Boolean).join('\n')

// ---- index ----------------------------------------------------------------------------------

/**
 * Every published record as a document. Nothing is filtered here: the bundle this is handed
 * was restricted to `published = 1` in SQL, so there is no second visibility rule to get wrong.
 */
export function buildIndex(content: Content, siteOrigin: string, resume?: string | null): Doc[] {
  const at = content.updatedAt
  const docs: Doc[] = []
  const src = (type: DocType, title: string, url = siteOrigin, extra: Partial<Source> = {}): Source => ({
    type,
    title,
    url,
    ...extra,
  })

  const site = content.site
  const links = content.socialLinks.filter((l) => l.url)
  docs.push({
    id: 'profile:about',
    type: 'profile',
    title: site.name || 'About',
    text: lines(
      site.name && `${site.name} — ${site.subtitle}`,
      ...site.paragraphs,
      site.email && `Email: ${site.email}`,
      ...links.map((l) => `${l.label}: ${l.url}`),
      site.resumeUrl && `Resume: ${site.resumeUrl}`,
    ),
    source: src('profile', site.name || 'About'),
    updatedAt: at,
  })

  for (const p of content.projects) {
    if (!p.slug) continue
    const url = `${siteOrigin}/projects/${p.slug}`
    const sections = sectionsOf(p.sections)
    const intro = p.sections.find((s) => 'text' in s.body)
    docs.push({
      id: `project:${p.slug}`,
      type: 'project',
      title: p.title,
      text: lines(
        p.tagline,
        p.status.label && `Status: ${p.status.label}`,
        p.stack.length > 0 && `Stack: ${p.stack.join(', ')}`,
        intro && sectionText(intro),
        sections.length > 0 && `Sections: ${sections.map((s) => s.slug).join(', ')}`,
        ...p.links.map((l) => `${l.label}: ${l.url}`),
        p.note,
        p.caveat && `Caveat: ${p.caveat}`,
      ),
      extra: [p.desktopLabel, ...(p.aliases ?? []), p.slug].join(' '),
      project: p.slug,
      source: src('project', p.title, url, { slug: p.slug }),
      updatedAt: at,
    })
    for (const s of sections) {
      docs.push({
        id: `research:${p.slug}/${s.slug}`,
        type: 'research',
        title: `${p.title} — ${s.heading}`,
        text: sectionText(s.section),
        extra: [p.desktopLabel, p.slug, ...(p.aliases ?? [])].join(' '),
        project: p.slug,
        section: s.slug,
        heading: s.heading,
        source: src('research', p.title, url, { slug: p.slug, section: s.slug }),
        updatedAt: at,
      })
    }
  }

  const keyed = (type: DocType, title: string, text: string, url = siteOrigin) => {
    const key = slugify(title) || 'item'
    // Two records with the same title must not collapse into one id.
    let id = `${type}:${key}`
    for (let n = 2; docs.some((d) => d.id === id); n++) id = `${type}:${key}-${n}`
    docs.push({ id, type, title, text, source: src(type, title, url), updatedAt: at })
  }
  for (const e of content.experience) keyed('experience', e.title, lines(e.title, e.detail, e.hint))
  for (const e of content.education) keyed('education', e.title, lines(e.title, e.detail, e.hint))
  for (const g of content.skills) keyed('skill', g.heading, `${g.heading}: ${g.items.join(', ')}`)
  for (const c of content.certificates) {
    keyed(
      'certificate',
      c.title,
      lines(
        `${c.title} — issued by ${c.issuer}`,
        c.issueDate && `Issued: ${c.issueDate}`,
        c.credentialUrl && `Credential: ${c.credentialUrl}`,
      ),
      c.credentialUrl ?? c.fileUrl ?? siteOrigin,
    )
  }

  for (const part of resumeSections(resume ?? '')) {
    docs.push({
      id: `resume:${part.slug}`,
      type: 'resume',
      title: `Resume — ${part.heading}`,
      text: part.text,
      section: part.slug,
      heading: part.heading,
      source: src('resume', 'Resume', content.site.resumeUrl, { section: part.slug }),
      updatedAt: at,
    })
  }

  return docs.filter((d) => d.text.trim())
}

/**
 * Split resume text on its own headings — the short all-capitals lines a resume is laid out
 * with (EDUCATION, TECHNICAL SKILLS, WORK EXPERIENCE…). Whatever comes before the first one is
 * the header block: name and contact line.
 */
export function resumeSections(text: string): { heading: string; slug: string; text: string }[] {
  const out: { heading: string; slug: string; lines: string[] }[] = []
  const lines = text.split(/\r?\n/).map((l) => l.trim()).filter(Boolean)
  lines.forEach((line, i) => {
    const heading = i > 0 && /^[A-Z][A-Z &/,-]{2,40}$/.test(line) && /[A-Z]{3}/.test(line)
    if (heading || !out.length) {
      const name = heading ? line.charAt(0) + line.slice(1).toLowerCase() : 'Contact'
      out.push({ heading: name, slug: slugify(name), lines: heading ? [] : [line] })
    } else out[out.length - 1]!.lines.push(line)
  })
  return out.map((s) => ({ heading: s.heading, slug: s.slug, text: s.lines.join('\n') })).filter((s) => s.text)
}

function sectionsOf(list: ProjectSection[]) {
  const seen = new Set<string>()
  return list.map((section, i) => {
    const heading = section.heading?.trim() || `Section ${i + 1}`
    let slug = slugify(heading) || `section-${i + 1}`
    if (seen.has(slug)) slug = `${slug}-${i + 1}`
    seen.add(slug)
    return { heading, slug, section }
  })
}

// ---- search ---------------------------------------------------------------------------------

export type SearchInput = { query: string; type?: DocType | 'all'; project?: string; limit?: number }

export type SearchResult = {
  id: string
  type: DocType
  title: string
  snippet: string
  source: Source
  updatedAt: string
}

/** Every query term, plus what it stands for when it names a kind of section. */
function expand(term: string): string[] {
  const out = [term]
  for (const [name, words] of Object.entries(SECTION_SYNONYMS)) {
    if (tokens(name).has(term)) for (const w of words) out.push(...tokens(w))
  }
  return out
}

export function search(index: Doc[], input: SearchInput): { results: SearchResult[]; message?: string } {
  const limit = Math.min(Math.max(1, input.limit ?? 5), MAX_RESULTS)
  const terms = [...tokens(input.query)].filter((t) => !STOPWORDS.has(t))
  const phrase = input.query.trim().toLowerCase()

  const pool = index.filter(
    (d) =>
      (!input.type || input.type === 'all' || d.type === input.type) &&
      (!input.project || d.project === input.project),
  )

  const scored = pool
    .map((doc) => {
      const bag = tokens(`${doc.title} ${doc.text} ${doc.extra ?? ''}`)
      const title = tokens(`${doc.title} ${doc.extra ?? ''}`)
      let hits = 0
      let score = 0
      const matched: string[] = []
      for (const term of terms) {
        const route = ROUTES.filter(([words]) => words.includes(term))
        const routed = route.some(([, types]) => types.includes(doc.type))
        const found = expand(term).find((t) => bag.has(t))
        if (found || routed) {
          hits++
          score += 1
          if (found) matched.push(found)
          // A category word ("projects") lifts its category; it is not a title match elsewhere,
          // or a certificate called "Structuring ML Projects" outranks every project.
          if (routed) score += 1
          else if (!route.length && title.has(term)) score += 1.5
        }
      }
      if (phrase.length > 3 && `${doc.title}\n${doc.text}`.toLowerCase().includes(phrase)) score += 3
      return { doc, hits, score, matched }
    })
    // A doc must answer most of the question, not one incidental word of it: "favorite movie"
    // must not come back as the movie-recommendation project.
    .filter((r) => terms.length > 0 && r.hits / terms.length > 0.5)
    .sort((a, b) => b.score - a.score || a.doc.id.localeCompare(b.doc.id))
    .slice(0, limit)

  const results = scored.map(({ doc, matched }) => ({
    id: doc.id,
    type: doc.type,
    title: doc.title,
    snippet: snippet(doc.text, matched),
    source: doc.source,
    updatedAt: doc.updatedAt,
  }))
  return results.length ? { results } : { results, message: NO_MATCH }
}

function snippet(text: string, matched: string[]): string {
  const flat = text.replace(/\s*\n\s*/g, ' · ').replace(/\s+/g, ' ').trim()
  if (flat.length <= SNIPPET_CHARS) return flat
  const lower = flat.toLowerCase()
  const at = Math.min(...matched.map((m) => lower.indexOf(m)).filter((i) => i >= 0), flat.length)
  const start = at === flat.length ? 0 : Math.max(0, at - 60)
  const cut = flat.slice(start, start + SNIPPET_CHARS - 2)
  return `${start > 0 ? '…' : ''}${cut}…`
}

// ---- retrieval ------------------------------------------------------------------------------

export type ContextResult =
  | {
      found: true
      id: string
      type: DocType
      title: string
      section?: string
      text: string
      truncated: boolean
      availableSections?: string[]
      source: Source
      updatedAt: string
    }
  | { found: false; message: string; availableSections?: string[] }

const cap = (text: string) =>
  text.length > MAX_TEXT_CHARS
    ? { text: `${text.slice(0, MAX_TEXT_CHARS)}…`, truncated: true }
    : { text, truncated: false }

/**
 * Resolve a caller-supplied section name against a project's real sections: exact slug first,
 * then the synonym table. Returns every match, in the project's own order.
 */
export function matchSections(docs: Doc[], wanted: string): Doc[] {
  const exact = docs.filter((d) => d.section === wanted)
  if (exact.length) return exact
  const words = SECTION_SYNONYMS[wanted] ?? [wanted.replace(/-/g, ' ')]
  return docs.filter((d) => {
    const heading = (d.heading ?? '').toLowerCase()
    return words.some((w) => heading.includes(w))
  })
}

export function getContext(index: Doc[], id: string, section?: string): ContextResult {
  const doc = index.find((d) => d.id === id)
  if (!doc) return { found: false, message: NO_MATCH }

  const sections = doc.type === 'project' ? index.filter((d) => d.type === 'research' && d.project === doc.project) : []
  const availableSections = sections.length ? sections.map((d) => d.section!) : undefined

  if (!section) return { found: true, ...pick(doc), ...cap(doc.text), availableSections }

  if (doc.type !== 'project' && doc.type !== 'research') {
    return { found: false, message: `${NO_MATCH} This record has no sections.` }
  }
  const pool = doc.type === 'project' ? sections : [doc]
  const hits = matchSections(pool, section)
  if (!hits.length) {
    return {
      found: false,
      message: `${NO_MATCH} "${section}" is not a section of ${doc.source.title}.`,
      availableSections: pool.map((d) => d.section!),
    }
  }
  const text = hits.map((d) => `## ${d.heading}\n${d.text}`).join('\n\n')
  const first = hits[0]!
  return {
    found: true,
    ...pick(first),
    title: doc.source.title,
    section: hits.map((d) => d.section).join(', '),
    ...cap(text),
    availableSections,
  }
}

const pick = (d: Doc) => ({ id: d.id, type: d.type, title: d.title, section: d.section, source: d.source, updatedAt: d.updatedAt })

// ---- compact views --------------------------------------------------------------------------

export function listProjects(content: Content, siteOrigin: string) {
  return {
    projects: content.projects
      .filter((p) => p.slug)
      .map((p) => ({
        id: `project:${p.slug}`,
        slug: p.slug,
        title: p.title,
        tagline: p.tagline,
        status: p.status.label,
        stack: p.stack,
        featured: p.featured,
        url: `${siteOrigin}/projects/${p.slug}`,
      })),
    updatedAt: content.updatedAt,
  }
}

export const PROFILE_SECTIONS = ['identity', 'skills', 'experience', 'education', 'certificates', 'links', 'resume'] as const
export type ProfileSection = (typeof PROFILE_SECTIONS)[number]

/**
 * The public profile, section by section. Only fields the site already shows a visitor; no
 * database ids, no file keys — the certificate id is a UUID and is deliberately left out.
 */
export function getProfile(content: Content, siteOrigin: string, section?: ProfileSection, resume?: string | null) {
  const s = content.site
  const all = {
    identity: { name: s.name, subtitle: s.subtitle, about: s.paragraphs, email: s.email, site: siteOrigin },
    skills: content.skills.map((g) => ({ group: g.heading, items: g.items })),
    experience: content.experience.map((e) => ({ title: e.title, detail: e.detail, when: e.hint })),
    education: content.education.map((e) => ({ title: e.title, detail: e.detail, when: e.hint })),
    certificates: content.certificates.map((c) => ({
      title: c.title,
      issuer: c.issuer,
      issued: c.issueDate || undefined,
      url: c.credentialUrl ?? c.fileUrl,
    })),
    links: content.socialLinks.map((l) => ({ label: l.label, handle: l.handle, url: l.url })),
    resume: {
      url: s.resumeUrl,
      sections: resumeSections(resume ?? '').map((r) => `resume:${r.slug}`),
      ...(section === 'resume' && resume ? cap(resume) : {}),
    },
  }
  const body: Partial<typeof all> = section ? { [section]: all[section] } : all
  return { ...body, source: { type: 'profile' as const, title: s.name, url: siteOrigin }, updatedAt: content.updatedAt }
}
